### PR TITLE
ShopifyAPI::Clients::HttpResponse as argument for ShopifyAPI::Errors::HttpResponseError

### PR DESCRIPTION
**Pull request by kaarelss**
ShopifyAPI::Clients::HttpResponse as argument for ShopifyAPI::Errors::HttpResponseError because code can be obtained from response as well as headers for custom retry logic when call limit is exceeded and 429 error occurs.

## Description

Fixes [#1038](https://github.com/Shopify/shopify-api-ruby/issues/1038)

Please, include a summary of what the PR is for:
- Ability to get headers from response when error occurs. Previously you can get only code from the error. 

## How has this been tested?

All tests passed as well as new test is included for this new functionality: `test_throttle_error_with_retry_after_header_in_error_object` in `http_client_test.rb`

## Checklist:

- [x] My commit message follow the pattern described in [here](https://chris.beams.io/posts/git-commit/)
- [x] I have performed a self-review of my own code.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the project documentation.
- [ ] I have added a changelog line.

### CODE REVIEW

**Review by jbourassa — APPROVED** [review]
I don't normally maintain this repo but this looks good to me. I'd like to see this merged if you can sign the CLA and fix the linter.

**Review by nelsonwittwer — CHANGES_REQUESTED** [review]

**Comment by jbourassa** [comment]
Sorbet complains about a missing type annotation, this should fix it:

```diff
diff --git lib/shopify_api/errors/http_response_error.rb lib/shopify_api/errors/http_response_error.rb
index 1d536a4..e4d912f 100644
--- lib/shopify_api/errors/http_response_error.rb
+++ lib/shopify_api/errors/http_response_error.rb
@@ -15,7 +15,7 @@ module ShopifyAPI
       sig { params(response: ShopifyAPI::Clients::HttpResponse).void }
       def initialize(response:)
         super
-        @code = response.code.to_i
+        @code = T.let(response.code, Integer)
         @response = response
       end
     end
```

You can run run the Sorbet locally with `bundle exec srb tc`.

Make sure to sign the CLA too: https://cla.shopify.com/

**Comment by kaarelss** [comment]
> Sorbet complains about a missing type annotation, this should fix it:
> 
> ```diff
> diff --git lib/shopify_api/errors/http_response_error.rb lib/shopify_api/errors/http_response_error.rb
> index 1d536a4..e4d912f 100644
> --- lib/shopify_api/errors/http_response_error.rb
> +++ lib/shopify_api/errors/http_response_error.rb
> @@ -15,7 +15,7 @@ module ShopifyAPI
>        sig { params(response: ShopifyAPI::Clients::HttpResponse).void }
>        def initialize(response:)
>          super
> -        @code = response.code.to_i
> +        @code = T.let(response.code, Integer)
>          @response = response
>        end
>      end
> ```
> 
> You can run run the Sorbet locally with `bundle exec srb tc`.
> 
> Make sure to sign the CLA too: https://cla.shopify.com/

Thank you @jbourassa for mentioning this, fixed it. We do not have sorbet in our ruby on rails project, so I'm new to this. About the CLA, my employer has asked Shopify to add additional authorized contributor to the existing Corporate CLA. We are still waiting for a reply from Shopify.

File: lib/shopify_api/clients/http_client.rb [hunk 1]
```diff
@@ -69,13 +69,13 @@ def request(request)
...
-            raise ShopifyAPI::Errors::HttpResponseError.new(code: response.code.to_i), error_message
+            raise ShopifyAPI::Errors::HttpResponseError.new(response: response), error_message
           end
 
           if tries == request.tries
-            raise ShopifyAPI::Errors::HttpResponseError.new(code: response.code), error_message if request.tries == 1
+            raise ShopifyAPI::Errors::HttpResponseError.new(response: response), error_message if request.tries == 1
 
-            raise ShopifyAPI::Errors::MaxHttpRetriesExceededError.new(code: response.code),
+            raise ShopifyAPI::Errors::MaxHttpRetriesExceededError.new(response: response),
```
[the remark this answers is not in the document]
Comment (kaarelss) [reply lib/shopify_api/clients/http_client.rb:78]: Added assertions for that - https://github.com/Shopify/shopify-api-ruby/pull/1040/commits/3312bc065d6a57fd2786fbfd4c80067599173c1f

**Comment by kaarelss** [comment]
> > HI @jbourassa and @nelsonwittwer. I have CLA signed now. What are the next steps?
> 
> @kaarelss can you rebase your branch with main? that should also get your tests to pass and we'll get this merged in

Done

**Comment by kaarelss** [comment]
> One last little 👶 step and we can get this merged in! Will you add a quick line in the change log describing your change?

Added

**Comment by hypogealgaol** [comment]
Hi, not sure if this is related but I wanted to try to create my own error doing the following: ```
response = ShopifyAPI::Clients::HttpResponse.new(code: 401, body: {}, headers: {})
ShopifyAPI::Errors::HttpResponseError.new(response: response)
Traceback (most recent call last):
        3: from (irb):175
        2: from (irb):176:in `rescue in irb_binding'
        1: from (irb):176:in `new'
ArgumentError (missing keyword: :code)
```

Am I doing something wrong?

**Pull request merged**